File: src/mrl/MysteriesRL.java
/*
 *  "MysteriesRL", Java Roguelike game.
 *  Copyright (C) 2015 by Serg V. Zhdanovskih (aka Alchemist, aka Norseman).
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package mrl;

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.awt.event.MouseMotionListener;
import javax.swing.JFrame;
import javax.swing.Timer;
import jzrlib.jterm.JTerminal;
import mrl.core.GlobalData;
import mrl.core.Locale;
import mrl.views.MainView;

/**
 *
 * @author Serg V. Zhdanovskih
 */
public final class MysteriesRL extends JFrame
{
    private final JTerminal fTerminal;

    private MainView fMainView;
    private Timer fTimer;
    private Locale fLocale;

    public MysteriesRL()
    {
        super();

        this.setResizable(false);
        this.fTerminal = new JTerminal(160, 80);
        this.fTerminal.setFocusable(true);
        this.fTerminal.setFocusTraversalKeysEnabled(false);
        this.add(fTerminal);
        this.pack();
        this.setLocationRelativeTo(null);
        this.setTitle(GlobalData.MRL_NAME);

        this.fLocale = new Locale();
        
        this.initViews();
        this.initTerminal();
        this.initTimer();
    }

    private void initViews()
    {
        this.fMainView = new MainView(this, null, this.fTerminal);
    }

    private void initTimer()
    {
        this.fTimer = new Timer(100, new ActionListener()
        {
            @Override
            public void actionPerformed(ActionEvent e)
            {
                fMainView.tick();
                repaint();

                if (JTerminal.DEBUG_PAINT) {
                    setTitle(String.valueOf(fTerminal.CharsPainted));
                }
            }
        });

        this.fTimer.setRepeats(true);
        this.fTimer.start();
    }

    private void initTerminal()
    {
        this.fTerminal.addKeyListener(new KeyListener()
        {
            @Override
            public void keyPressed(KeyEvent e)
            {
                fMainView.keyPressed(e);
            }

            @Override
            public void keyReleased(KeyEvent e)
            {
                //System.out.println("test");
            }

            @Override
            public void keyTyped(KeyEvent e)
            {
                fMainView.keyTyped(e);
            }
        });

        this.fTerminal.addMouseMotionListener(new MouseMotionListener()
        {
            @Override
            public void mouseDragged(MouseEvent e)
            {
            }

            @Override
            public void mouseMoved(MouseEvent e)
            {
                fMainView.mouseMoved(e);
            }
        });

        this.fTerminal.addMouseListener(new MouseListener()
        {
            @Override
            public void mouseClicked(MouseEvent e)
            {
                fMainView.mouseClicked(e);
            }

            @Override
            public void mousePressed(MouseEvent e)
            {
            }

            @Override
            public void mouseReleased(MouseEvent e)
            {
            }

            @Override
            public void mouseEntered(MouseEvent e)
            {
            }

            @Override
            public void mouseExited(MouseEvent e)
            {
            }
        });
    }

    public static void main(String[] args)
    {
        MysteriesRL app = new MysteriesRL();
        app.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        app.setVisible(true);
    }
}

File: project/MysteriesRL.cs
/*
 *  "MysteriesRL", roguelike game.
 *  Copyright (C) 2015 by Serg V. Zhdanovskih (aka Alchemist, aka Norseman).
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

using System.IO;
using System.Reflection;
using System.Timers;
using System.Windows.Forms;
using ZRLib.Core;
using MysteriesRL.Core;
using MysteriesRL.Views;
using ZRLib.Terminal;

namespace MysteriesRL
{
    public sealed class MRLWin : Form
    {
        private readonly Terminal fTerminal;

        private Locale fLocale;
        private MainView fMainView;
        private System.Timers.Timer fTimer;

        public MRLWin()
            : base()
        {
            fTerminal = new Terminal(160, 80);
            fTerminal.KeyPress += KeyTyped;
            fTerminal.KeyUp += KeyPressed;
            fTerminal.MouseDown += MouseClicked;
            fTerminal.MouseMove += MouseMoved;
            fTerminal.Dock = DockStyle.Fill;

            ClientSize = fTerminal.Size;
            Controls.Add(fTerminal);
            MaximizeBox = false;
            MinimizeBox = true;
            Text = MRLData.MRL_NAME;
            FormBorderStyle = FormBorderStyle.FixedSingle;
            StartPosition = FormStartPosition.CenterScreen;

            Logger.LogInit(GetAppPath() + "MysteriesRL.log");

            fLocale = new Locale();

            fMainView = new MainView(null, fTerminal);

            fTimer = new System.Timers.Timer(250);
            fTimer.Elapsed += TickTimer;
            fTimer.Start();
        }

        private void TickTimer(object sender, ElapsedEventArgs e)
        {
            fMainView.Tick();
            fTerminal.Invalidate(false);
        }

        public void KeyPressed(object sender, KeyEventArgs e)
        {
            fMainView.KeyPressed(e);
        }

        public void KeyTyped(object sender, KeyPressEventArgs e)
        {
            fMainView.KeyTyped(e);
        }

        public void MouseMoved(object sender, MouseEventArgs e)
        {
            fMainView.MouseMoved(e);
        }

        public void MouseClicked(object sender, MouseEventArgs e)
        {
            fMainView.MouseClicked(e);
        }

        public static string GetAppPath()
        {
            Module[] mods = Assembly.GetExecutingAssembly().GetModules();
            string fn = mods[0].FullyQualifiedName;
            return Path.GetDirectoryName(fn) + Path.DirectorySeparatorChar;
        }
    }
}
